climada_petals/hazard/copernicus_forecast/handler.py

```diff
@@ -0,0 +1,560 @@
+"""
+This file is part of CLIMADA.
+
+Copyright (C) 2017 ETH Zurich, CLIMADA contributors listed in AUTHORS.
+
+CLIMADA is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free
+Software Foundation, version 3.
+
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Module to handle seasonal forecast data from the Copernicus Climate Data Store (CDS)
+in the U-CLIMADAPT project.
+
+TODO: update prerequisites
+Prerequisites:
+1. CDS API client installation:
+   pip install cdsapi
+
+2. CDS account and API key:
+   Register at https://cds-beta.climate.copernicus.eu
+
+3. CDS API configuration:
+   Create a .cdsapirc file in your home directory with your API key and URL.
+   For instructions, visit:
+   https://cds-beta.climate.copernicus.eu/how-to-api#install-the-cds-api-client
+"""
+
+import os
+import logging
+import calendar
+import re
+
+import xarray as xr
+import pandas as pd
+import numpy as np
+import cdsapi
+
+from climada.hazard import Hazard
+from climada.util.coordinates import get_country_geometries
+import climada_petals.hazard.copernicus_forecast.indicator as indicator
+
+
+LOGGER = logging.getLogger(__name__)
+
+class ForecastHandler:
+    """
+    A class to handle downloading, processing, and calculating climate indices
+    and hazards based on seasonal forecast data from Copernicus Climate Data Store (CDS).
+    """
+
+    _FORMAT_GRIB = 'grib'
+    _FORMAT_NC = 'nc'
+    
+    def __init__(self, data_dir='.', url = None, key = None):
+        """
+        Initializes the ForecastHandler instance.
+
+        Parameters:
+        data_dir (str): Path to the directory where downloaded and processed data will be stored.
+            Defaults to the current directory ('.').
+        url (str): url to the CDS API. Defaults to None, in which case the url from /.cdsapirc
+            is used.
```

Comment:
Please use NumpyDoc style for all docstrings.

Reply:
Done